File: 백준/Gold/1644. 소수의 연속합/소수의 연속합.cc
#include <bits/stdc++.h>

using namespace std;

int n;
int arr[4000001];
vector<int> v;


// 에라토스테네스의 체
/**
* 2는 소수이다. 그러면 2의 배수들은 전부 소수가 아니다
* 3은 소수이다. 그러면 3의 배수들은 전부 소수가 아니다
* ...
* 소수가 아닌 수만 남게됨.
*/

void chae(int n)
{
	fill(arr, arr + 4000001, 1);

	arr[1] = 0;

	for (int i = 2; i <= n; i++)
	{
		if (arr[i] == 1) // 해당 수가 소수라면
		{
			v.push_back(i);
			for (int j = i * 2; j <= n; j += i) // 해당 수의 배수들은 전부 소수가 아님
			{
				arr[j] = 0;
			}
		}
	}
}

int main()
{
	cin >> n;

	chae(n);

	int st = 0;
	int en = 0;
	int sum = 0;
	int result = 0;

	while (1)
	{
		if (sum > n)
		{
			sum -= v[st++];
		}
		else if (sum < n)
		{
			if (en >= v.size())
				break;

			sum += v[en++];
		}
		else
		{
			result++;

			if (en >= v.size())
				break;

			sum += v[en];
			en++;
		}
	}

	cout << result;

	return 0;
}

File: 백준/Gold/1644. 소수의 연속합/소수의 연속합.java
import java.util.*;
import java.io.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static StringBuilder sb = new StringBuilder();

    static int n, count;
    static boolean[] prime;
    static Integer[] arr;
    static ArrayList<Integer> list;

    public static void main(String[] args) throws Exception {
        n = Integer.parseInt(br.readLine());

        prime = new boolean[n+1];
        list = new ArrayList<>();

        Arrays.fill(prime, true);

        prime[1] = false;

        for (int i = 2 ; i <= n ; i ++) {
            if (prime[i]) {
                int j = 2;
                while(i * j <= n) {
                    prime[i*j] = false;
                    j++;
                }
            }
        }

        for (int i = 1 ; i <= n ; i ++) {
            if (prime[i])
                list.add(i);
        }

        arr = list.toArray(new Integer[0]);

        int left = 0;
        int right = 0;
        int sum = 0;

        while(true) {
            // 목표 값 보다 작으면 right를 한 칸 밀기
            // 목표 값 보다 크면 left를 한 칸 밀기
            // 목표 값이면 count + 1;

            if (sum == n) {
                count ++;
                sum -= arr[left++];
            } else if (sum < n) {
                if (right >= arr.length) break;
                sum += arr[right++];
            } else if (sum > n) {
                sum -= arr[left++];
            }
        }

        System.out.println(count);
    }
}